### PR TITLE
Made scrollbars in trees perfect again.

### LINKED ISSUE #2637: Perfect scrollbars don't work in widgets which are using react-virtualized (all trees)
Since react-virtualized is creating an own inner scrollcontainer the perfect scrollbar does not work anymore in widgets with trees (and all other widgets which use react-virtualized). 
So, the perfect scrollbar has to be related to the react-virtualized container somehow.

### PR DESCRIPTION
**Pull request by jbicker**
Fixes #2637

### CODE REVIEW

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 1]
```diff
@@ -241,12 +241,17 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
...
                 renderNodeRow={this.renderNodeRow}
                 scrollToRow={this.scrollToRow}
+                handleScroll={this.handleScroll}
             />;
         }
         // tslint:disable-next-line:no-null-keyword
         return null;
     }
 
+    protected readonly handleScroll = (info: { clientHeight: number; scrollHeight: number; scrollTop: number }) => {
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-widget.tsx]: `(params: ScrollParams) => {` where `ScrollParams` comes from `react-virtualized`

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 2]
```diff
@@ -566,6 +571,12 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
         this.addKeyListener(this.node, up, event => this.handleUp(event));
         this.addKeyListener(this.node, down, event => this.handleDown(event));
         this.addKeyListener(this.node, Key.ENTER, event => this.handleEnter(event));
+        this.node.addEventListener('ps-scroll-y', (e: any) => {
+            if (this.view && this.view.list && this.view.list.Grid) {
+                const { scrollTop } = e.target;
+                this.view.list.Grid.handleScrollEvent({ scrollLeft: 0, scrollTop });
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-widget.tsx]: `scrollLeft:  0` was just to try, we should remove it

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 3]
```diff
@@ -566,6 +571,12 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
         this.addKeyListener(this.node, up, event => this.handleUp(event));
         this.addKeyListener(this.node, down, event => this.handleDown(event));
         this.addKeyListener(this.node, Key.ENTER, event => this.handleEnter(event));
+        this.node.addEventListener('ps-scroll-y', (e: any) => {
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-widget.tsx]: `e: { target: { scrollTop: number } }`

File: packages/core/src/browser/tree/tree-widget.tsx [hunk 4]
```diff
@@ -752,13 +763,14 @@ export namespace TreeWidget {
         height: number
         scrollToRow?: number
         rows: NodeRow[]
+        handleScroll: (info: { clientHeight: number; scrollHeight: number; scrollTop: number }) => void
```
Comment (akosyakov) [inline packages/core/src/browser/tree/tree-widget.tsx]: `handleScroll: (param: ScrollParams) => void`

**Comment by kittaakos** [comment]
I am reviewing it now... Should it work in FF as well?

(I have restarted the Travis build.)

**Comment by akosyakov** [comment]
@kittaakos yes, it is the main point to make it look nice if FF :)

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 3 above)
Comment (jbicker) [reply packages/core/src/browser/tree/tree-widget.tsx]: That does not work. The compiler complains.
```
[typescript]
Argument of type '(e: { target: { scrollTop: number; }; }) => void' is not assignable to parameter of type 'EventListenerOrEventListenerObject'.
  Type '(e: { target: { scrollTop: number; }; }) => void' is not assignable to type 'EventListenerObject'.
    Property 'handleEvent' is missing in type '(e: { target: { scrollTop: number; }; }) => void'.
```

**Comment by kittaakos** [comment]
I have a lot of errors in the backend log, not sure if this is related:
```
root ERROR ../../packages/search-in-workspace/lib/browser/search-in-workspace-result-tree-widget.js/SearchInWorkspaceResultTreeWidget.prototype.updateCurrentEditorDecorations/<@http://localhost:3000/59.bundle.js:1057:21
../../packages/search-in-workspace/lib/browser/search-in-workspace-result-tree-widget.js/SearchInWorkspaceResultTreeWidget.prototype.updateCurrentEditorDecorations@http://localhost:3000/59.bundle.js:1053:9
../../packages/search-in-workspace/lib/browser/search-in-workspace-result-tree-widget.js/SearchInWorkspaceResultTreeWidget.prototype.init/<@http://localhost:3000/59.bundle.js:914:13
../../packages/core/lib/common/event.js/CallbackList.prototype.invoke@http://localhost:3000/bundle.js:98871:26
../../packages/core/lib/common/event.js/Emitter.prototype.fire@http://localhost:3000/bundle.js:98951:13
../../packages/editor/lib/browser/editor-manager.js/EditorManager.prototype.setActiveEditor@http://localhost:3000/11.bundle.js:621:13
../../packages/editor/lib/browser/editor-manager.js/EditorManager.prototype.updateActiveEditor@http://localhost:3000/11.bundle.js:626:9
../../packages/editor/lib/browser/editor-manager.js/EditorManager.prototype.init/<@http://localhost:3000/11.bundle.js:603:63
invokeSlot@http://localhost:3000/bundle.js:8302:13
emit@http://localhost:3000/bundle.js:8260:17
../../node_modules/@phosphor/signaling/lib/index.js/</Signal.prototype.emit@http://localhost:3000/bundle.js:7933:9
../../packages/core/lib/browser/shell/application-shell.js/ApplicationShell.prototype.onActiveChanged@http://localhost:3000/bundle.js:90335:9
invokeSlot@http://localhost:3000/bundle.js:8302:13
emit@http://localhost:3000/bundle.js:8260:17
../../node_modules/@phosphor/signaling/lib/index.js/</Signal.prototype.emit@http://localhost:3000/bundle.js:7933:9
../../node_modules/@phosphor/widgets/lib/focustracker.js/FocusTracker.prototype._setWidgets@http://localhost:3000/bundle.js:14466:13
../../node_modules/@phosphor/widgets/lib/focustracker.js/FocusTracker.prototype._evtFocus@http://localhost:3000/bundle.js:14480:9
../../node_modules/@phosphor/widgets/lib/focustracker.js/FocusTracker.prototype.handleEvent@http://localhost:3000/bundle.js:14443:17
TextAreaWrapper</TextAreaWrapper.prototype.setSelectionRange@http://localhost:3000/vs/editor/editor.main.js:29218:17
TextAreaState</TextAreaState.prototype.writeToTextArea@http://localhost:3000/vs/editor/editor.main.js:28502:17
TextAreaInput</TextAreaInput.prototype._setAndWriteTextAreaState@http://localhost:3000/vs/editor/editor.main.js:29085:13
TextAreaInput</TextAreaInput.prototype.writeScreenReaderContent@http://localhost:3000/vs/editor/editor.main.js:29093:13
TextAreaInput</TextAreaInput.prototype._setHasFocus@http://localhost:3000/vs/editor/editor.main.js:29071:21
TextAreaInput</TextAreaInput.prototype.focusTextArea@http://localhost:3000/vs/editor/editor.main.js:29053:13
TextAreaHandler</TextAreaHandler.prototype.focusTextArea@http://localhost:3000/vs/editor/editor.main.js:90175:13
View</View.prototype.focus@http://localhost:3000/vs/editor/editor.main.js:91446:13
CodeEditorWidget</CodeEditorWidget.prototype.focus@http://localhost:3000/vs/editor/editor.main.js:91619:13
../../packages/monaco/lib/browser/monaco-editor.js/MonacoEditor.prototype.focus@http://localhost:3000/16.bundle.js:230:9
../../packages/editor/lib/browser/editor-widget.js/EditorWidget.prototype.onActivateRequest@http://localhost:3000/11.bundle.js:1734:9
../../node_modules/@phosphor/widgets/lib/widget.js/Widget.prototype.processMessage@http://localhost:3000/bundle.js:23580:17
invokeHandler@http://localhost:3000/bundle.js:7518:13
sendMessage@http://localhost:3000/bundle.js:7254:13
runMessageLoop@http://localhost:3000/bundle.js:7568:17
```

**Comment by kittaakos** [comment]
> I have a lot of errors in the backend log, not sure if this is related:

This happens only with FF but not with Chrome.

**Comment by jbicker** [comment]
@kittaakos, let me check.

File: packages/core/src/browser/tree/tree-widget.tsx (hunk 3 above)
Comment (kittaakos) [reply packages/core/src/browser/tree/tree-widget.tsx]: I assume this listener should be removed later, right?

**Comment by kittaakos** [comment]
Please check the build on Linux, it has failed twice in a row, while the OS X ones passed.

**Comment by kittaakos** [comment]
I have verified on Windows and OS X both in Chrome and FF. It looks great!

Please note, I did not check the code so thoroughly. Please remove the [listeners](
https://github.com/theia-ide/theia/pull/2724#discussion_r214263211), and fix the [Linux build](
https://github.com/theia-ide/theia/pull/2724#issuecomment-417577259). (Maybe it is unrelated to your change and you just have to clear the caches on Travis.)

**Comment by jbicker** [comment]
Regarding the listeners we have already filed an issue: https://github.com/theia-ide/theia/issues/2254
It's mainly navigation related but I will get rid of any further listeners here as well.

**Comment by kittaakos** [comment]
Please squash the commits next time.

**Pull request merged**